gnomad_qc/v4/create_release/calculate_variant_statistics.py

```diff
@@ -24,188 +51,218 @@
-            metric_expr, [0.0, 0.25, 0.5, 0.75, 1.0]
+    if by_ancestry and meta_ht is None:
+        raise ValueError(
+            "If `by_ancestry` is True, a Table containing sample metadata is required."
         )
 
-        return hl.struct(mean=metric_mean, quantiles=metric_quantiles)
-
-    def _get_agg_expr(ht):
-        return hl.struct(
+    agg_expr = {
+        strat: hl.struct(
             **{
-                strat: hl.struct(
-                    **{
-                        _metric: _get_metric_stats(ht, strat, _metric)
-                        for _metric in qc_metrics
-                    }
+                metric: hl.struct(
+                    mean=hl.agg.mean(ht[strat][metric]),
+                    quantiles=hl.agg.approx_quantiles(
+                        ht[strat][metric], [0.0, 0.25, 0.5, 0.75, 1.0]
+                    ),
                 )
-                for strat in strats
+                for metric in ht[strat]
+                if isinstance(ht[strat][metric], hl.expr.NumericExpression)
             }
         )
+        for strat in ht.row_value
+        if isinstance(ht[strat], hl.expr.StructExpression)
+    }
 
-    ht = ht.annotate_globals(all_samples=ht.aggregate(_get_agg_expr(ht)))
+    stats_struct = ht.aggregate(agg_expr, _localize=False)
 
     if by_ancestry:
-        meta_ht = meta(data_type=data_type).ht()
-        ht = ht.annotate(gen_anc=meta_ht[ht.s].population_inference.pop)
-        by_ancestry_agg = ht.aggregate(hl.agg.group_by(ht.gen_anc, _get_agg_expr(ht)))
-        # Remove 'None' key, otherwise it can't be annotated to globals
-        by_ancestry_agg.pop(None)
-        ht = ht.annotate_globals(**by_ancestry_agg)
+        # Performing aggregation for all samples and by genetic ancestry.
+        # Note: This is performing multiple aggregations, rather than a single group_by
+        # aggregation to avoid a Class too large error.
+        gen_anc_expr = meta_ht.population_inference.pop
+        gen_ancs = meta_ht.aggregate(
+            hl.agg.filter(
+                hl.is_defined(gen_anc_expr), hl.agg.collect_as_set(gen_anc_expr)
+            )
+        )
+        gen_anc_expr = meta_ht[ht.s].population_inference.pop
+
+        stats_struct = hl.struct(
+            all_samples=stats_struct,
+            stratified_by_ancestry=hl.struct(
+                **{
+                    anc: ht.aggregate(
+                        hl.agg.filter(gen_anc_expr == anc, agg_expr), _localize=False
+                    )
+                    for anc in gen_ancs
+                }
+            ),
+        )
 
-    return ht
+    return stats_struct
 
 
 def main(args):
-    """Collect summary stats of gnomAD v4 data."""
+    """Collect per-sample variant counts and aggregate sample statistics."""
     hl.init(
-        log="/per_exome.log",
+        log="/per_sample_stats",
         default_reference="GRCh38",
         tmp_dir="gs://gnomad-tmp-30day",
     )
```

Comment:
```suggestion
        tmp_dir="gs://gnomad-tmp-4day",
```

Reply:
yeah, I went back and forth on this one actually. I'm a little worried about how long this will run and don't want to have a temp file removed mid run if it happens to run for a long time